crates/bevy_ecs/src/storage/sparse_set.rs

```diff
@@ -258,10 +266,11 @@ impl<I: SparseSetIndex, V> SparseSet<I, V> {
         if let Some(dense_index) = self.sparse.get(index.clone()).cloned() {
             // SAFE: dense indices stored in self.sparse always exist
             unsafe {
-                *self.dense.get_unchecked_mut(dense_index) = value;
+                *self.dense.get_unchecked_mut(dense_index.get()) = value;
             }
         } else {
-            self.sparse.insert(index.clone(), self.dense.len());
+            self.sparse
+                .insert(index.clone(), NonMaxUsize::new(self.dense.len()).unwrap());
```

Comment:
This isn't an uncommon branch to take, and this introduces a potential panic, even if we never panic normally, this will have a negative impact on performance, and it's not something we can use `debug_checked_unwrap` on either since we know for a fact that it's something that we can hit.

Reply:
Does that mean you want me to revert that line?

Reply:
I honestly am not sure how to deal with this. This branch is unavoidable if we want to avoid unsoundness. If we can find a way to address this without impacting performance, this seems feasible, but otherwise, this is potentially on the hotpath for a lot of operations within the engine, so any tangible performance regression is probably unacceptable. Definitely need to benchmark this in some way to validate.

Reply:
What is the potential performance issue here? Is it `xor` operation or `unwrap()`?

Reply:
The best way to deal with this is to improve Rust language (there're a lot of discussions like [this](https://internals.rust-lang.org/t/elevate-nonzero-types-to-true-primitive-integer-types/18040)). But I'm afraid we're not getting it in the next ten years. :(

Reply:
> What is the potential performance issue here? Is it xor operation or unwrap()?

The unwrap, the XOR has persistently shown to be easily pipelinable in any current CPUs without adding too much here, but the extra branch and codegen from panicking has shown to have a signficant impact when working with code that is this hot.